lib/Extractor/KLEEBuilder.cpp

```diff
@@ -169,6 +172,23 @@ ref<Expr> ExprBuilder::makeSizedArrayRead(unsigned Width, StringRef Name,
                                                                                   klee::ConstantExpr::create(i, Width))));
       PowerTwoBitsMap[Origin] = PowerExpr;
     }
+    if (Origin->NumSignBits > 1) {
+      ref<Expr> MSB = ExtractExpr::create(Var, Width-1, Expr::Bool);
+      ref<Expr> SignBitsExpr = klee::ConstantExpr::create(1, Expr::Bool);
+      for (int I=Width-1; I >= (int) (Width-Origin->NumSignBits); --I)
+        SignBitsExpr = AndExpr::create(SignBitsExpr, EqExpr::create(MSB, ExtractExpr::create(Var, I, Expr::Bool)));
+      if (Origin->NumSignBits != Width) {
+         int J = (int) Width - 1- Origin->NumSignBits;
+         ref<Expr> Expr1 = ExtractExpr::create(Var, J, Expr::Bool);
+         ref<Expr> Expr2 = ExtractExpr::create(Var, J+1, Expr::Bool);
+         SignBitsExpr = AndExpr::create(SignBitsExpr, EqExpr::create(AndExpr::create(Expr1, Expr2),
+                                        klee::ConstantExpr::create(0, Expr::Bool)));
+      }
+      SignBitsExpr = AndExpr::create(SignBitsExpr,
+                                     Expr::createImplies(EqExpr::create(MSB, klee::ConstantExpr::create(0, Expr::Bool)),
+                                                         EqExpr::create(ctlz(Var), klee::ConstantExpr::create(Origin->NumSignBits, Width))));
```

Comment:
Yes, the idea of asserting all bits are either 0 or 1 is much more efficient than the for loop. I think if we add extra cases to teach souper about number of sign bits. Like, I have added the cases to handle the variables which are both power of two and have N sign bits. Should I keep this as a simple expression or can I add more cases to make it more better in terms of finding more optimizations?